lib/postscript/request.ex

```diff
@@ -45,19 +45,47 @@ defmodule Postscript.Request do
 
     request
     |> config.http_client.send(config.http_client_opts)
+    |> maybe_validate_json_decode(config)
     |> retry(request, config)
     |> finish(config)
   end
 
-  defp retry(response, _request, %_{ retry: retry }) when is_nil(retry) or retry == false do
+  defp maybe_validate_json_decode({:ok, %{body: body, headers: headers}} = response, config) do
+    headers
+    |> Enum.reduce(%{}, fn {k, v}, acc -> Map.put(acc, String.downcase(k), v) end)
+    |> case do
+      %{"content-type" => "application/json"} ->
+        case config.json_codec.decode(body) do
+          {:ok, _decoded} ->
+            response
+
+          {:error, decode_error} ->
+            Logger.warning([
+              inspect(__MODULE__),
+              " received an invalid JSON response ",
+              inspect(body)
+            ])
+
+            {:error, decode_error}
+        end
+
+      _otherwise ->
+        response
+    end
+  end
+
+  defp maybe_validate_json_decode(response, _config), do: response
```

Comment:
I feel like we're loading several assumptions into the code here. I think starting with the changes you made in `Postscript.Response` is a good place to start. If there's something more specific that needs to be garnered from the response I think the application is the right place to do that.

Reply:
Agreed